src/Middleware/HeaderPropagation/src/DependencyInjection/HeaderPropagationExtensions.cs

```diff
@@ -0,0 +1,38 @@
+using System;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.HeaderPropagation
+{
+    public static class HeaderPropagationExtensions
+    {
+        public static IServiceCollection AddHeaderPropagation(this IServiceCollection services, Action<HeaderPropagationOptions> configure)
+        {
+            services.TryAddSingleton<HeaderPropagationState>();
+            services.Configure(configure);
+
+            return services;
+        }
+
+        public static IHttpClientBuilder AddHeaderPropagation(this IHttpClientBuilder builder)
+        {
+            builder.Services.TryAddSingleton<HeaderPropagationState>();
+            builder.Services.TryAddTransient<HeaderPropagationMessageHandler>();
+ 
+            builder.AddHttpMessageHandler(services =>
+            {
+                var state = services.GetRequiredService<HeaderPropagationState>();
+                return new HeaderPropagationMessageHandler(services.GetRequiredService<IOptions<HeaderPropagationOptions>>(), state);
+            });
+
+            return builder;
+        }
+
+        public static IApplicationBuilder UseHeaderPropagation(this IApplicationBuilder builder)
+        {
+            return builder.UseMiddleware<HeaderPropagationMiddleware>();
```

Comment:
This should check for `HeaderPropagationState` to be registered in services and throw a friendly error reminding the user to call `AddHeaderPropagation`. This is something we're starting to do for all middleware to make mistakes more figure-out-able.

Reply:
I've copyed the message from https://github.com/aspnet/AspNetCore/blob/85e2147ff037cc9950249fbb9d095ad47ec4184a/src/Http/Routing/src/Resources.resx#L199.
Is there a better approach?

Reply:
Loooks good